osquery/tables/system/posix/ssh_keys.cpp

```diff
@@ -16,10 +16,31 @@
 #include <osquery/tables/system/system_utils.h>
 #include <osquery/utils/system/system.h>
 
+#include <boost/algorithm/string.hpp>
+
 namespace osquery {
 namespace tables {
 
 const std::string kSSHUserKeysDir = ".ssh/";
+const std::string kEd25519Header = "-----BEGIN OPENSSH PRIVATE KEY-----";
+
+// The first bytes of an OpenSSH key are |key type|length of cipher name|cipher
+// name| so all unencrypted ed25519 keys should start with the value below
+// (encoded in base64) This magic string is
+//  6f 70 65 6e 73 73 68 2d 6b 65 79 2d 76 31 00 00 00 00 04 6e 6f 6e 65
+// |--------------------------------------------|-----------|----------|
+// | o  p  e  n  s  s  h  -  k  e  y  -  v  1 \0|          4| n  o  n e|
+//
+const std::string kEd25519UnencryptedPrefix = "b3BlbnNzaC1rZXktdjEAAAAABG5vbmU";
+
+bool isEncrypted(std::string keys_content) {
+  if (boost::starts_with(keys_content, kEd25519Header)) {
+    const std::string prefix = keys_content.substr(
+        kEd25519Header.size() + 1, kEd25519UnencryptedPrefix.size());
+    return prefix != kEd25519UnencryptedPrefix;
```

Comment:
auto keys_body = keys_content.substr(kEd25519Header.size() + 1)
return boost::starts_with(keys_body, kEd25519UnencryptedPrefix)

Reply:
Thanks for resolving the issue.

I'm afraid substr(
        kEd25519Header.size() + 1, kEd25519UnencryptedPrefix.size()) can crash the code if string is smaller. Does my suggestion look okay?

Reply:
Whoops, it's been a long time since I've written any C-esque code.

I'll push up a fix momentarily.

Reply:
Ah, actually `.substr` does do what I thought – it just gets characters to the end of the string if the second argument is greater than the length. However, this will still crash if the string is exactly `kEd25519Header` since substr throws if the first argument is greater. So your suggestion made me notice a bug even though it doesn't quite work!